File: functional/0-high-order-function/main.go
package main

import "fmt"

type words []string
type wordsFiltered func(words) words
type wordCondition func(string) bool

func main() {
	groupA := words{"the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog"}

	conditionA := anyCondition([]wordCondition{
		isSameWordAt("over"),
		haveOddCharacter,
	})

	conditionB := allCondition([]wordCondition{
		haveAlphabet,
		haveEvenCharacter,
	})

	wordFilterABy := wordFilter(conditionA)
	wordFilterBBy := wordFilter(conditionB)

	fmt.Printf("Condition A: %+v\n", wordFilterABy(groupA))
	fmt.Printf("Condition B: %+v\n", wordFilterBBy(groupA))
}

func anyCondition(conditions []wordCondition) wordCondition {
	return func(word string) bool {
		for _, c := range conditions {
			if c(word) {
				return true
			}
		}
		return false
	}
}

func allCondition(conditions []wordCondition) wordCondition {
	return func(word string) bool {
		for _, c := range conditions {
			if !c(word) {
				return false
			}
		}
		return true
	}
}

func wordFilter(condition wordCondition) wordsFiltered {
	return func(ws words) words {
		var newWords words
		for _, word := range ws {
			if condition(word) {
				newWords = append(newWords, word)
			}
		}
		return newWords
	}
}

func haveAlphabet(word string) bool {
	for _, c := range word {
		if c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' {
			return true
		}
	}
	return false
}

func haveEvenCharacter(word string) bool {
	return len(word)%2 == 0
}

func haveOddCharacter(word string) bool {
	return len(word)%2 != 0
}

func isSameWordAt(wordCondition string) wordCondition {
	return func(word string) bool {
		return wordCondition == word
	}
}

File: functional/0-high-order-function/main.dart
typedef List<String> WordsFiltered(List<String> words);
typedef bool WordCondition(String string);

void main() {
  List<String> groupA = [
    "the",
    "quick",
    "brown",
    "fox",
    "jumps",
    "over",
    "the",
    "lazy",
    "dog"
  ];

  WordCondition conditionA = anyCondition(<WordCondition>[
		isSameWordAt("over"),
		haveOddCharacter,
  ]);

	WordCondition conditionB = allCondition(<WordCondition>[
		haveAlphabet,
		haveEvenCharacter,
  ]);

	WordsFiltered wordFilterABy = wordFilter(conditionA);
	WordsFiltered wordFilterBBy = wordFilter(conditionB);

	print("Condition A: ${wordFilterABy(groupA)}\n");
	print("Condition B: ${wordFilterBBy(groupA)}\n");
}

WordCondition anyCondition(List<WordCondition> conditions) {
    bool condition(String word) {
      for (WordCondition condition in conditions) {
        if (condition(word)) return true;
      }
      return false;
    }

    return condition;
  }

  WordCondition allCondition(List<WordCondition> conditions) {
    bool condition(String word) {
      for (WordCondition condition in conditions) {
        if (!condition(word)) return false;
      }
      return true;
    }

    return condition;
  }

  WordsFiltered wordFilter(WordCondition condition) {
    List<String> filtering(List<String> words) {
      List<String> filtered = [];
      for (String word in words) {
        if (condition(word)) filtered.add(word);
      }
      return filtered;
    }

    return filtering;
  }

  bool haveAlphabet(String word) {
    for (String c in word.split("")) {
      if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u') {
        return true;
      }
    }
    return false;
  }

  bool haveEvenCharacter(String word) {
    return word.length % 2 == 0;
  }

  bool haveOddCharacter(String word) {
    return word.length % 2 != 0;
  }

  WordCondition isSameWordAt(String inputWord) {
    bool condition(String word) {
      return inputWord == word;
    }

    return condition;
  }
